src/utils/manager.py

```diff
@@ -107,156 +162,46 @@ def supported(self) -> bool:
-
-        Args:
-            mountpoint: share mountpoint to unmount.
+        old_autofs = self._autofs_file.read_text()
 
-        Raises:
-            Error: Raised if the unmount operation fails.
-        """
-        _logger.debug(f"Unmounting share at mountpoint {mountpoint}")
-        autofs_id = _mountpoint_to_autofs_id(mountpoint)
-        pathlib.Path(f"/etc/auto.{autofs_id}").unlink(missing_ok=True)
-        pathlib.Path(f"/etc/auto.master.d/{autofs_id}.autofs").unlink(missing_ok=True)
+        # Avoid restarting autofs if the config didn't change.
+        if not force_mount and new_autofs == old_autofs:
+            return
 
         try:
+            for mount in mounts._mounts.keys():
+                pathlib.Path(mount).mkdir(parents=True, exist_ok=True)
+            self._autofs_file.write_text(new_autofs)
             systemd.service_reload("autofs", restart_on_failure=True)
         except systemd.SystemdError as e:
-            _logger.error(f"Failed to unmount {mountpoint}. Reason:\n{e}")
-            raise Error(f"Failed to unmount {mountpoint}")
-
-        shutil.rmtree(mountpoint, ignore_errors=True)
-
-
-def _trigger_autofs() -> None:
-    """Triggers a mount on all filesystems handled by autofs.
-
-    This function is useful to make autofs-managed mounts appear on the
-    `/proc/mount` file, since they could be unmounted when reading the file.
-    """
-    for fs in _mounts("autofs"):
-        _logger.info(f"triggering automount for `{fs.mountpoint}`")
-        try:
-            os.scandir(fs.mountpoint).close()
-        except OSError as e:
-            # Not critical since it could also be caused by unrelated mounts,
-            # but should be good to log it in case this causes problems.
-            _logger.warning(f"Could not trigger automount for `{fs.mountpoint}`. Reason:\n{e}")
-
-
-def _mountpoint_to_autofs_id(mountpoint: Union[str, os.PathLike]) -> str:
-    """Get the autofs id of a mountpoint path.
-
-    Args:
-        mountpoint: share mountpoint.
-    """
-    path = pathlib.Path(mountpoint).resolve()
-    return str(path).lstrip("/").replace("/", "-")
-
-
-def _mounts(fstype: str = "") -> Iterator[MountInfo]:
-    """Get an iterator of all mounts in the system that have the requested fstype.
-
-    Returns:
-        Iterator[MountInfo]: All the mounts with a valid fstype.
-    """
-    with pathlib.Path("/proc/mounts").open("rt") as mounts:
-        for mount in mounts:
-            # Lines in /proc/mounts follow the standard format
-            # <endpoint> <mountpoint> <fstype> <options> <freq> <passno>
-            m = MountInfo(*mount.split())
-            if fstype and not m.fstype.startswith(fstype):
-                continue
-
-            yield m
+            _logger.error(f"failed to mount filesystems. reason:\n{e}")
+            if "Operation not permitted" in str(e) and not self.supported():
+                raise Error("mounting shares not supported on LXD containers")
```

Comment:
Should the charm become blocked at this state rather than just raising an error/failing hook execution? I would expect that we should be checking if we're in a container before we get here (e.g. during install hook). If not, we should do so (can be done in separate PR).